In [1]:
import csv
import xlrd
import numpy as np
import pandas as pd

# https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )


In [3]:
annual_geographical_mobility_rates_type_movement
# annual_geographical_mobility_rates_type_movement.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  \
0    Table A-1. Annual Geographical Mobility Rates,...                           
1                              (Numbers in thousands.)                           
2                                                  NaN                           
3                                      Mobility period                           
4                                                  NaN                           
5                                                  NaN                           
6                                               NUMBER                           
7                                                  NaN                           
8                                            2017-2018                           
9                                            2016-2017                           
10                                           2015-2016                           
11                                                 NaN                           
12                                           2014-2015                           
13                   2013-2014 (98,000 address file)/1                           
14                                           2012-2013                           
15                                           2011-2012                           
16                         2010-2011 (2010 controls)/2                           
17                         2010-2011 (2000 controls)/3                           
18                                                 NaN                           
19                         2009-2010 (2010 controls)/2                           
20                         2009-2010 (2000 controls)/3                           
21                                           2008-2009                           
22                                           2007-2008                           
23                                           2006-2007                           
24                                           2005-2006                           
25                                                 NaN                           
26                                         2004-2005/4                           
27                                           2003-2004                           
28                                           2002-2003                           
29                                           2001-2002                           
..                                                 ...                           
156                                          1960-1961                           
157                                                NaN                           
158                                          1959-1960                           
159                                          1958-1959                           
160                                          1957-1958                           
161                                          1956-1957                           
162                                          1955-1956                           
163                                                NaN                           
164                                          1954-1955                           
165                                          1953-1954                           
166                                          1952-1953                           
167                                          1951-1952                           
168                                          1950-1951                           
169                                                NaN                           
170                                          1949-1950                           
171                                          1948-1949                           
172                                          1947-1948              

annual_geographical_mobility_rates_type_movement

notes:
    seems to cover back to 1947
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [3]:
# annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample(3)

Table with row headers in column A and column headers in row 5. Unnamed: 1  \
15                            .Net internal migration                   -200*   
7                                        .Outmigrants                     595   
69                                        .Inmigrants                     248   

   Unnamed: 2 Unnamed: 3 Unnamed: 4  
15        -25         44       181*  
7         555        714        689  
69        415        899        617

annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions

notes:
    seems to cover back to 1980
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [4]:
# inmigration_outmigration__net_migration_metropolitan_status
inmigration_outmigration__net_migration_metropolitan_status.sample(3)

Table with row headers in column A and column headers in row 5.  \
210  Note: Data for 1985-1986 and 1986-1987 are cor...                
101                                          2001-2002                
65                              .Nonmetropolitan areas                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
210        NaN        NaN        NaN        NaN        NaN  
101        NaN        NaN        NaN        NaN        NaN  
65         974       1159      -185*         61       -124

inmigration_outmigration__net_migration_metropolitan_status

notes:
    seems to cover back to 1985
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [5]:
# geographical_mobility_by_tenure
geographical_mobility_by_tenure.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
132                                                NaN                                                                                                                                 
407                            .Total, 1 year and over                                                                                                                                 
437                            .Total, 1 year and over                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
132        NaN        NaN        NaN        NaN        NaN        NaN   
407        100        100        100        100        100        100   
437        100        100        100        100        100        100   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
132        NaN        NaN        NaN  
407        100        100        100  
437        100        100        100

geographical_mobility_by_tenure

notes:
    seems to cover back to 1987
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [6]:
# reason_for_move_specific_categories
reason_for_move_specific_categories.sample(3)

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
49                                          2005-2006                                                                                    
46                                          2008-2009                                                                                    
12                                          2012-2013                                                                                    

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
49        100      6.012    8.50717    13.1561    8.73811    1.60153   
46        100    5.37124    9.47042    11.4971    8.74276    2.70853   
12      35918       1817       3753       5301       3242        750   

   Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14  \
49    3.61473   0.426739    4.01386  ...     17.7975     4.40294     6.15257   
46    5.04514   0.398868     1.0376  ...     14.5183     5.04245     11.1171   
12       1941        237        809  ...        5332        1135        2989   

   Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \
49           -     9.23513     2.67088     0.43929     1.26516     1.67934   
46           -     9.72915     2.57378    0.528231     1.59547    0.363832   
12         654        5016         215          20         136          11   

   Unnamed: 21  
49     1.71449  
46     4.78103  
12         462  

[3 rows x 22 columns]

reason_for_move_specific_categories

notes:
    seems to cover back to 1998
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [7]:
# distance_of_intercounty_move
distance_of_intercounty_move.sample(3)

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
25                                        2003-2004/4                                                                     
26                                        2002-2003/4                                                                     
42                        2009-2010 (2000 controls)/3                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
25      15171       4665       3722       3186       3598  
26      15356       4953       3407       3177       3819  
42        100    39.2172    21.3104    14.1817    25.2907

distance_of_intercounty_move

notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

## Aggregate Dates
    - first recording year varies by up to 55 years 
| first recording year | dataset |
|------|------|
| 1947 | annual_geographical_mobility_rates_type_movement |
|------|------|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|------|------|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1998 | reason_for_move_specific_categories |
|------|------|
| 2002 | distance_of_intercounty_move |

- let's drop the rows clustered with NaNs

In [8]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
170                                          1949-1950                           
176  3/ Population controls consistent with 2000 Ce...                           
100                        2010-2011 (2010 controls)/2                           
60                                         1970-1971/8                           
0    Table A-1. Annual Geographical Mobility Rates,...                           
122                                          1995-1996                           
159                                          1958-1959                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
170        100       80.9       19.1       18.7       13.1        5.6   
176        NaN        NaN        NaN        NaN        NaN        NaN   
100        100       88.4       11.6       11.2        7.7        3.5   
60      201506     163800      37705      36161      23018      13143   
0          NaN        NaN        NaN        NaN        NaN        NaN   
122        100       83.7       16.3       15.8       10.3        5.6   
159        100       80.3       19.7       19.2       13.1        6.1   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
170          3        2.6        0.3  
176        NaN        NaN        NaN  
100        1.9        1.6        0.4  
60        6197       6946       1544  
0          NaN        NaN        NaN  
122        3.1        2.5        0.5  
159        3.2          3        0.5

notes:
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value

actions:
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [9]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# reset column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.sample(7)

Mobility period Total, 1 year old and over  \
89   2000-2001 (1990 controls)/6                        100   
50                     1964-1965                     187974   
79                     2008-2009                        100   
81                     2006-2007                        100   
135                    1947-1948                        100   
127                    1955-1956                        100   
7    2010-2011 (2010 controls)/2                     302640   

    Same residence (non-movers) Total movers  \
89                      85.8283      14.1713   
50                       149128        38846   
79                         87.5         12.5   
81                         86.8         13.2   
135                        79.8         20.2   
127                        78.9         21.1   
7                        267602        35038   

    Different residence in the United States Same county Different county  \
89                                   13.5669     7.98875          5.57815   
50                                     37866       25122            12744   
79                                      12.1         8.4              3.7   
81                                      12.8         8.6              4.2   
135                                     19.9        13.6              6.4   
127                                     20.5        13.7              6.8   
7                                      33953       23330            10623   

    Same state Different state Movers from abroad  
89     2.76194         2.81621           0.604391  
50        6597            6147                978  
79         2.1             1.6                0.4  
81         2.5             1.7                0.4  
135        3.3             3.1                0.3  
127        3.6             3.1                0.6  
7         5868            4756               1084

as of now, we have:
    1) removed rows with NaN values
    2) extracted, sorted, and set column names
    
next:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            2017-2018
            2013-2014 (98,000 address file)/1
            2011-2012
            2010-2011 (2010 controls)/2
            2000-2001 (SCHIP, 2000 controls)/5
            1970-1971/8
        - what we want:
            2018 << 2017-2018
            2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            2012 << 2011-2012
            2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            1971 << 1970-1971/8

In [10]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

0

as test_000 returns 0:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity 
    
next up:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [11]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

total: 110
ratio: 0.8088235294117647


notes:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson
    
actions:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [12]:
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1
# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

119                           1964
98     1992-1993 (1980 controls)/7
76     2010-2011 (2000 controls)/3
117                           1966
131                           1952
46                            1969
8      2010-2011 (2000 controls)/3
Name: Mobility period, dtype: object

- notes
    - we have converted the low hanging fruit

- actions
    - now on to exploring the other 26 rows

In [23]:
# print them out , let's see what we can see
_ = 0
for row in df['Mobility period']:
    _ += 1    
    if not isinstance( row , int ):
        if _ < 10:
            print( str(_) + '  |  ' + row )
        elif _ < 100:
            print( str(_) + ' |  ' + row )
        else:
            print( str(_) + '|  ' + row )

5  |  2013-2014 (98,000 address file)/1
8  |  2010-2011 (2010 controls)/2
9  |  2010-2011 (2000 controls)/3
10 |  2009-2010 (2010 controls)/2
11 |  2009-2010 (2000 controls)/3
16 |  2004-2005/4
20 |  2000-2001 (2000 controls)/3
21 |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2000-2001 (1990 controls)/6
30 |  1992-1993 (1990 controls)/6
31 |  1992-1993 (1980 controls)/7
44 |  1975-1976/8
45 |  1970-1971/8
73 |  2013-2014 (98,000 address file)/1
76 |  2010-2011 (2010 controls)/2
77 |  2010-2011 (2000 controls)/3
78 |  2009-2010 (2010 controls)/2
79 |  2009-2010 (2000 controls)/3
84 |  2004-2005/4
88 |  2000-2001 (2000 controls)/3
89 |  2000-2001 (SCHIP, 2000 controls)/5
90 |  2000-2001 (1990 controls)/6
98 |  1992-1993 (1990 controls)/6
99 |  1992-1993 (1980 controls)/7
112|  1975-1976/8
113|  1970-1971/8


- notes
    - 6 rows formatted almost like majority (new lowest)
        - all have either /4 or /8 at the end of otherwise standard string
            - e.g. standard 
                - 2004-2005
            - observation rows 
                - 16 |  2004-2005/4
                - 44 |  1975-1976/8
                - 45 |  1970-1971/8
                - 84 |  2004-2005/4
                - 112|  1975-1976/8
                - 113|  1970-1971/8
        - remarks before reading documentation
            - identical instances
                - rows 16 and 84
                - rows 44 and 112
                - rows 45 and 113
            - rows are not ordered by year
                - unless years reoccour/loop through table
                - or otherwise
    - 16 rows are in the next new lowest format
        - all contain ({year} controls)/{num} with year and num in consistent pairs
            - (year / num) pairs = # occurrences
                - (2010 / 2) = 4
                - (2000 / 3) = 6
                - (1990 / 6) = 4
                - (1980 / 7) = 2
        - they occour in pattern twice:
            - 4 uninterrupted
            - 1
            - 3 uninterrupted
        - sample rows
            - 8  |  2010-2011 (2010 controls)/2
            - 9  |  2010-2011 (2000 controls)/3
            - 10 |  2009-2010 (2010 controls)/2
            - 11 |  2009-2010 (2000 controls)/3
        - unknown analysis
            - seems to be game settings for different censuses 
            - difficult to connect the /2,/3 and /6,/7 relationship
                - if was dependent on decade order , possibles would be straight ordered 
                    - e.g. /2, /3, /4, /5 or /1, /3, /5, /7
                    
    - final 4 rows are longer and are 2 duplicate pairs
        - default format + ending either (98,000 address file)/1 or (SCHIP, 2000 controls)/5
            - longest rows
                - 5  |  2013-2014 (98,000 address file)/1
                - 21 |  2000-2001 (SCHIP, 2000 controls)/5
                - 73 |  2013-2014 (98,000 address file)/1
                - 89 |  2000-2001 (SCHIP, 2000 controls)/5
        - pre document remarks
            - first notice of /1 
            - return use of /5 
                - do other /n have additional possibile (paramaters)
    - blind analysis
        - year range
                - min: 1970-1971 (1971 in complete df)
                - max: 2013-2014 (2014 in complete df)
            - less range than full representation of column
                - min: 1948 (1947-1948 in origional df)
                - max: 2018 (2017-2018 in origional df)
            - why?
                - recent changes to census
                    - collection
                    - questions
                    - coverage
                - recent increase in census related surveys
                    - housing surveys
                    - business surveys
                        - we will be exploring these after fitting gen_000 model
                        - these are also how this data is presented for non-decade years
        - lot of possible meanings for non-conforming objects
                - politics..
                - changing standards
                - technological evolution
                - etc...
                    - storage , collection , methods , etc... 
            - full-court lobs
                - something to do with immigrants
                - evolution of process 
                - expansion of non-physical opportunities 
        - there's got to be a better way to extract dataframes from excel files
            - likely convertable, but annotations likely direct either universe to same result 
